feat(agent): Add Yii v2 framework instrumentation

razvanphp opened the pull request:
- [x] Implement Yii v2 instrumentation for auto transaction naming
- [x] Implement non-web transactions (yiic CLI) — works out-of-the-box
- [x] Update documentation
- [x] Implement wrapper for [Yii2 ErrorHandler](https://www.yiiframework.com/doc/api/2.0/yii-web-errorhandler) (currently NewRelic is blind because of the custom PHP error_handler of the framework)
- [ ] Add integration tests (needed?)

Fixes #821

I've recompiled the agent locally with those changes and transaction names correctly show up now:

![Screenshot 2024-01-31 at 07 23 53](https://github.com/newrelic/newrelic-php-agent/assets/4599319/ce01be96-e102-45f0-8721-05c505556ae3)

![Screenshot 2024-01-31 at 11 32 02](https://github.com/newrelic/newrelic-php-agent/assets/4599319/350f73b5-1b88-430c-bd1e-808d7f529fc0)

CLI uses the same hook method:

![Screenshot 2024-01-31 at 11 16 34](https://github.com/newrelic/newrelic-php-agent/assets/4599319/9505ff44-cab9-4eec-aa23-3855ae438480)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/newrelic/newrelic-php-agent?pullRequest=823) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/newrelic/newrelic-php-agent?pullRequest=823) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/newrelic/newrelic-php-agent?pullRequest=823) it.</sub>

lavarou commented:
@razvanphp Thank you for engaging and your interest in improving New Relic’s PHP instrumentation! Could you update this PR and set `dev` as the base branch? Here’s a picture how to do it after you click on the ‘Edit’ button next to PR’s title:
<img width="1254" alt="Screenshot 2024-02-01 at 3 52 13 PM" src="https://github.com/newrelic/newrelic-php-agent/assets/98839644/bd1416d8-7f41-41a5-b4fd-94277c260f6d">

razvanphp commented:
hey @lavarou thanks for taking a look here. I've changed the base branch and also rebased my `origin/main` branch on `dev`. Can you please help me also implement the last two things on my list? Namely the ErrorHandler and the integration tests? I don't have much experience with that...

lavarou commented:
> Can you please help me also implement the last things on my list? Namely the integration tests? I don't have much experience with that...

@razvanphp, thank you for your contribution! Please take a look at an example integration test for WordPress: [tests/integration/frameworks/wordpress/test_wordpress_00.php](https://github.com/newrelic/newrelic-php-agent/blob/320ea571a11bc469d7d8179dfe51577b54df11df/tests/integration/frameworks/wordpress/test_wordpress_00.php). Per the [test description](https://github.com/newrelic/newrelic-php-agent/blob/320ea571a11bc469d7d8179dfe51577b54df11df/tests/integration/frameworks/wordpress/test_wordpress_00.php#L7-L15), it checks that:
- the framework is detected: [here](https://github.com/newrelic/newrelic-php-agent/blob/320ea571a11bc469d7d8179dfe51577b54df11df/tests/integration/frameworks/wordpress/test_wordpress_00.php#L32)
- transactions are named correctly: [here](https://github.com/newrelic/newrelic-php-agent/blob/320ea571a11bc469d7d8179dfe51577b54df11df/tests/integration/frameworks/wordpress/test_wordpress_00.php#L33) 
- no errors are generated: [here](https://github.com/newrelic/newrelic-php-agent/blob/320ea571a11bc469d7d8179dfe51577b54df11df/tests/integration/frameworks/wordpress/test_wordpress_00.php#L50) 

It is using a mock of WordPress and the new integration tests can use mock as well. However, if installing the Yii2 framework is easier than creating a mock, it is possible to download and install additional files at the time the test executes, see [here](https://github.com/newrelic/newrelic-php-agent/blob/320ea571a11bc469d7d8179dfe51577b54df11df/tests/integration/external/drupal7/skipif.inc#L13). 

You can use [containerized development environment](https://github.com/newrelic/newrelic-php-agent/blob/320ea571a11bc469d7d8179dfe51577b54df11df/docs/dev_environment.md#dockerized-development-environment-for-the-new-relic-php-agent) to build the agent and run the tests locally, only the New Relic license key is required.
